xds/internal/resolver/xds_resolver.go

```diff
@@ -88,26 +87,65 @@ func (b *xdsResolverBuilder) Build(target resolver.Target, cc resolver.ClientCon
+	// - updates from the xDS client. This could lead to generation of new
+	//   service config if resolution is complete.
+	// - completion of an RPC to a removed cluster causing the associated ref
+	//   count to become zero, resulting in generation of new service config.
+	// - stopping of a config selector that results in generation of new service
+	//   config.
+	ctx, cancel := context.WithCancel(context.Background())
+	r.serializer = grpcsync.NewCallbackSerializer(ctx)
+	r.serializerCancel = cancel
+
+	// Initialize the xDS client.
 	newXDSClient := rinternal.NewXDSClient.(func() (xdsclient.XDSClient, func(), error))
 	if b.newXDSClient != nil {
 		newXDSClient = b.newXDSClient
 	}
-
 	client, close, err := newXDSClient()
 	if err != nil {
 		return nil, fmt.Errorf("xds: failed to create xds-client: %v", err)
 	}
 	r.xdsClient = client
 	r.xdsClientClose = close
+
+	// Determine the listener resource name and start a watcher for it.
+	template, err := r.sanityChecksOnBootstrapConfig(target, opts, r.xdsClient)
+	if err != nil {
+		return nil, err
+	}
+	endpoint := target.URL.Path
+	if endpoint == "" {
+		endpoint = target.URL.Opaque
+	}
+	endpoint = strings.TrimPrefix(endpoint, "/")
+	r.ldsResourceName = bootstrap.PopulateResourceTemplate(template, endpoint)
+	r.listenerWatcher = newListenerWatcher(r.ldsResourceName, r)
+	return r, nil
+}
+
+// Performs the following sanity checks:
+//   - Verifies that the bootstrap configuration is not empty.
+//   - Verifies that the bootstrap configuration contains certificate providers if
+//     the use of xDS credentials is specified by the user.
+//   - Verifies that if the provided dial target contains an authority, the
+//     bootstrap configuration contains server config for that authority.
+//
+// Returns the listener resource name template to use. If any of the above
+// validations fail, a non-nil error is returned.
+func (r *xdsResolver) sanityChecksOnBootstrapConfig(target resolver.Target, opts resolver.BuildOptions, client xdsclient.XDSClient) (string, error) {
 	bootstrapConfig := client.BootstrapConfig()
 	if bootstrapConfig == nil {
-		return nil, errors.New("bootstrap configuration is empty")
+		// This is never expected to happen after a successful xDS client
+		// creation. Defensive programming.
+		return "", fmt.Errorf("xds: bootstrap configuration is empty")
 	}
 
-	// If xds credentials were specified by the user, but bootstrap configs do
-	// not contain any certificate provider configuration, it is better to fail
-	// right now rather than failing when attempting to create certificate
-	// providers after receiving an CDS response with security configuration.
+	// If xDS credentials were specified by the user, but the bootstrap config
+	// does not contain any certificate providers, it is better to fail right
+	// now rather than failing when attempting to create certificate providers
+	// after receiving an CDS response with security configuration.
```

Comment:
Yeah, I saw that. But it's also going through a gRFC process. So, we can worry about it when the gRFC gets merged.

Also, this check will be removed soon. See: https://github.com/grpc/grpc-go/issues/6756.

I don't want to fold that check into this PR because tests will need to be updated as well.

Reply:
Sounds good.